### PR TITLE
feat(Core/Hooks): Added new hooks on talents reset

### LINKED ISSUE #4327: Core/Talents:m_extraBonusTalentCount Only in CLASS_DEATH_KNIGHT
CURRENT BEHAVIOUR:
I'm a novice and I don't know much about GitHub, so I'll feed back the questions here. When I tested it, I found that only death knight can take effect when calculating additional talent points. The specific code involved is: \src\server\game\Entities\Player\Player.cpp Player::CalculateTalentsPoints()

Fix Code:
```
uint32 Player::CalculateTalentsPoints() const
{
    uint32 talentPointsForLevel = 0;
    uint32 base_talent = getLevel() < 10 ? 0 : getLevel() - 9;

    if (getClass() != CLASS_DEATH_KNIGHT || GetMapId() != 609)
    {
        talentPointsForLevel = base_talent;
    }
    else
    {
        talentPointsForLevel = getLevel() < 56 ? 0 : getLevel() - 55;
        talentPointsForLevel += m_questRewardTalentCount;
        if (talentPointsForLevel > base_talent)
            talentPointsForLevel = base_talent;
    }

    talentPointsForLevel += m_extraBonusTalentCount;
    return uint32(talentPointsForLevel * sWorld->getRate(RATE_TALENT));
}
```

AC HASH/COMMIT:
[ab8967c](https://github.com/azerothcore/azerothcore-wotlk/commit/ab8967c6388684db95f1ea26a8c0692d1e8db941)

OPERATING SYSTEM:
Windows 10 20H2

MODULES:
NO MODULES

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/96067452-core-talents-m_extrabonustalentcount-only-in-class_death_knight?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by kaytotes**
## Changes Proposed:
-  Adds a hook within `InitTalentForLevel`. The actual hook already existed however wasn't fired.
-  Adds an `OnAfterTalentsReset` hook that fires once a players talent points have been successfully removed. There is an existing `OnPlayerTalentsReset` which is essentially the "before" and this is the "after". I have not renamed this hook to the logical `OnBeforePlayerTalentsReset` as I figured it could be breaking since that touches Eluna too. Maybe worth another PR with a clear point in the changelog?
- Pulled in the fix from attached issue as I also needed this for my module.

## Issues Addressed:
- Closes #4327

## Tests Performed:
- Docker build
- Custom player script in module that grants them 5 talent points every 15 levels so that at 60 they have 71. (See Below)

```c++
void OverrideTalents::OnBeforeInitTalentForLevel(Player* player, uint8& /*level*/ uint32& talentPointsForLevel)
{
    sLog->outError("Adjusting Talents for %s", player->GetName());

    uint8 level = player->getLevel();
    uint32 base_talents = level < 10 ? 0 : level - 9;
    uint32 extra_talents = 0;

    if (level >= 15)
        extra_talents += 5;

    if (level >= 30)
        extra_talents += 5;

    if (level >= 45)
        extra_talents += 5;

    if (level >= 60)
        extra_talents += 5;

    talentPointsForLevel = base_talents + extra_talents;
}

void OverrideTalents::OnAfterTalentsReset(Player* player)
{
    sLog->outError("Talents were reset for %s", player->GetName());

    player->InitTalentForLevel();
}
```

## How to Test the Changes:

1.
2.
3.

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/Entities/Player/Player.cpp [hunk 1]
```diff
@@ -12648,18 +12652,23 @@ void Player::StoreLootItem(uint8 lootSlot, Loot* loot)
...
-    if (talentPointsForLevel > base_talent)
+    {
         talentPointsForLevel = base_talent;
+    }
+    else
+    {
+        talentPointsForLevel = getLevel() < 56 ? 0 : getLevel() - 55;
+        talentPointsForLevel += m_questRewardTalentCount;
+        if (talentPointsForLevel > base_talent)
+            talentPointsForLevel = base_talent;
```
Comment (Helias) [inline src/server/game/Entities/Player/Player.cpp:12667]:
```suggestion
        if (talentPointsForLevel > base_talent)
        {
            talentPointsForLevel = base_talent;
        }
```

File: src/server/game/Entities/Player/Player.cpp [hunk 2]
```diff
@@ -12648,18 +12652,23 @@ void Player::StoreLootItem(uint8 lootSlot, Loot* loot)
...
-    uint32 talentPointsForLevel = getLevel() < 56 ? 0 : getLevel() - 55;
-    talentPointsForLevel += m_questRewardTalentCount;
-
-    if (talentPointsForLevel > base_talent)
+    {
         talentPointsForLevel = base_talent;
+    }
+    else
+    {
+        talentPointsForLevel = getLevel() < 56 ? 0 : getLevel() - 55;
```
Comment (Helias) [inline src/server/game/Entities/Player/Player.cpp:12664]: I suggest to assign getLevel() to a variable and do not call it twice.
```suggestion
        auto playerLevel = getLevel();
        talentPointsForLevel = playerLevel < 56 ? 0 : playerLevel - 55;
```

Comment (Helias) [reply src/server/game/Entities/Player/Player.cpp:12664]: looking better at the code, you can create `auto playerLevel = getLevel();` at the begin of this function and avoid calling getLevel() multiple times.

**Comment by Helias** [comment]
@kaytotes thanks for this PR, good job :rocket: 

Please, update also the documentation, have a look here https://www.azerothcore.org/wiki/hooks-script#documenting-your-hook

**Comment by kaytotes** [comment]
I'll add a changelog and fix the build today.

**Comment by kaytotes** [comment]
@Helias I've investigated the build error and had to revert the removal of the `level` variable on `OnBeforeInitTalentForLevel` to put it back. The `mod-azerothshard` module uses this hook and would need to be updated too. I may need to open a PR there in future or with this PR depending on what we want to do. 

Before this PR the function in the module couldn't have been firing so no idea just how relevant it is or how much that module is used.

I dislike that in the current implementation of this function we pass the players level by reference and then do basically nothing with that info.

**Comment by Helias** [comment]
You can change something if you consider it useless or not efficient

Mod-azerothshard is not used so much so far

**Comment by kaytotes** [comment]
In that case I've opened a matching PR for this one. No idea what order we have to merge things in. Probably there first https://github.com/azerothcore/mod-azerothshard/pull/10

File: src/server/game/Scripting/ScriptMgr.h [hunk 3]
```diff
@@ -908,7 +915,7 @@ class PlayerScript : public ScriptObject
     virtual void OnBeforeUpdateAttackPowerAndDamage(Player* /*player*/, float& /*level*/, float& /*val2*/, bool /*ranged*/) { }
     virtual void OnAfterUpdateAttackPowerAndDamage(Player* /*player*/, float& /*level*/, float& /*base_attPower*/, float& /*attPowerMod*/, float& /*attPowerMultiplier*/, bool /*ranged*/) { }
 
-    virtual void OnBeforeInitTalentForLevel(Player* /*player*/, uint8& /*level*/, uint32& /*talentPointsForLevel*/) { }
+    virtual void OnBeforeInitTalentForLevel(Player* /*player*/, uint32& /*talentPointsForLevel*/) { }
```
Comment (Nefertumm) [inline src/server/game/Scripting/ScriptMgr.h:918]: Also, add this change to the changelog... Something like, for hook OnBefore.... it doesn't require the parameter level anymore.

**Review by Nefertumm — APPROVED** [review]

**Comment by Nefertumm** [comment]
For me it looks great (just add what I marked above). It would be awesome if some tester can do the proper here too.

File: src/server/game/Entities/Player/Player.cpp [hunk 4]
```diff
@@ -12648,18 +12652,26 @@ void Player::StoreLootItem(uint8 lootSlot, Loot* loot)
...
+    }
+    else
+    {
+        talentPointsForLevel = currentLevel < 56 ? 0 : currentLevel - 55;
+        talentPointsForLevel += m_questRewardTalentCount;
+        if (talentPointsForLevel > baseTalents)
+        {
+            talentPointsForLevel = baseTalents;
+        }
+    }
```
Comment (FrancescoBorzi) [inline src/server/game/Entities/Player/Player.cpp:12671]: @kaytotes what is this logic change here for?

Comment (kaytotes) [reply src/server/game/Entities/Player/Player.cpp:12671]: Resolves https://github.com/azerothcore/azerothcore-wotlk/issues/4327 and allows custom content to easily award extra talents.

Comment (Winfidonarleyan) [reply src/server/game/Entities/Player/Player.cpp:12671]: You can close this pr if you don't keep working on it

**Comment by Winfidonarleyan** [comment]
If you don't keep working on it

**Pull request closed without merging**